### PR TITLE
feat: add @@auth option for declaring auth model

### PR DESCRIPTION
**Pull request by jasonmacdonald**
This is by no means complete yet, but I want to check that I'm on the right track. I wasn't able to find a simple way to pull the attribute value during the linker phase, so I added two utility functions - but there's probably a better way.

I'll need to figure out how to test in VScode Extension as well, as this is my first time delving into that.

### CODE REVIEW

File: packages/schema/src/language-server/zmodel-linker.ts [hunk 1]
```diff
@@ -278,9 +278,11 @@ export class ZModelLinker extends DefaultLinker {
                 const model = getContainingModel(node);
 
                 if (model) {
-                    const userModel = getAllDeclarationsFromImports(this.langiumDocuments(), model).find(
-                        (d) => isDataModel(d) && d.name === 'User'
-                    );
+                    const dataModel = this.getContainingDataModel(node);
```
Comment (ymc9) [inline packages/schema/src/language-server/zmodel-linker.ts:281]: I think we need to find data model with `@@auth` attribute from all data model declarations, but not just the container of the current `auth()` invocation. Just like how we used to find the model named "User" in the code below this.

File: packages/schema/src/res/stdlib.zmodel [hunk 2]
```diff
@@ -376,6 +376,11 @@ attribute @@deny(_ operation: String, _ condition: Boolean)
  */
 attribute @deny(_ operation: String, _ condition: Boolean)
 
+/**
+ * Defines the model to use when when checking access policies.
+ */
+attribute @@auth(_ modelName: String)
```
Comment (ymc9) [inline packages/schema/src/res/stdlib.zmodel:382]: Shall we just declare this attribute without a parameter? My thought is to use it like the following:

```prisma
model Foo {
    ...
    @@auth()
}
```

File: packages/schema/src/language-server/zmodel-linker.ts [hunk 3]
```diff
@@ -278,9 +278,11 @@ export class ZModelLinker extends DefaultLinker {
...
+                    const customAuthModel = dataModel && getModelAttributeValue('@@auth', dataModel);
```
Comment (ymc9) [inline packages/schema/src/language-server/zmodel-linker.ts:282]: If we declare the attribute without a parameter, we can simply find it by checking attribute name. There's a `hasAttribute` helper from `@zenstackhq/sdk` for doing this.

**Review by ymc9 — CHANGES_REQUESTED** [review]
Hi @jasonmacdonald , thanks a lot for making this PR. I've added some comments. Please let me know if you have time to continue working on it, and if not, I'll be happy to take it over and make tunings as needed.

For testing, I think you can check if `zenstack generate` works properly. If so, the VSCode extension will just work (since the CLI runs more check phases than the VSCode extension). The simplest way to test the CLI is to run `pnpm build` in "schema" folder, and copy over the content of "schema/dist" to "node_modules/zenstack" of an existing project.

File: packages/schema/src/language-server/zmodel-linker.ts (hunk 1 above)
Comment (jasonmacdonald) [reply packages/schema/src/language-server/zmodel-linker.ts:281]: ok, let me try that

File: packages/schema/src/res/stdlib.zmodel (hunk 2 above)
Comment (jasonmacdonald) [reply packages/schema/src/res/stdlib.zmodel:382]: Ah, so you were thinking the attribute would declare what Model is the "User" model. Ok, for some reason, I thought you wanted it to be declared in each place it was used, but that makes more sense.

**Comment by jasonmacdonald** [comment]
> Hi @jasonmacdonald , thanks a lot for making this PR. I've added some comments. Please let me know if you have time to continue working on it, and if not, I'll be happy to take it over and make tunings as needed.
> 
> For testing, I think you can check if `zenstack generate` works properly. If so, the VSCode extension will just work (since the CLI runs more check phases than the VSCode extension). The simplest way to test the CLI is to run `pnpm build` in "schema" folder, and copy over the content of "schema/dist" to "node_modules/zenstack" of an existing project.

I'll take another stab at it. I thought I might be going in the wrong direction, hence posting this early. I think I understand your thoughts on it now.

**Comment by jasonmacdonald** [comment]
@ymc9 Changed the attribute to be placed on the target model.

**Comment by ymc9** [comment]
> @ymc9 Changed the attribute to be placed on the target model.

Thank you, @jasonmacdonald ! It looks good to me now. I'll merge it soon and include it in the upcoming v1.2.0 release.

**Review by ymc9 — DISMISSED** [review]

**Review by ymc9 — APPROVED** [review]

**Comment by ymc9** [comment]
Fixes #774

**Pull request merged**